torch/nn/modules/activation.py

```diff
@@ -691,27 +691,39 @@ class MultiheadAttention(Module):
     Args:
         embed_dim: total dimension of the model.
         num_heads: parallel attention heads.
+        kdim: total number of features in key. Default: None.
+        vdim: total number of features in key. Default: None.
         dropout: a Dropout layer on attn_output_weights. Default: 0.0.
         bias: add bias as module parameter. Default: True.
         add_bias_kv: add bias to the key and value sequences at dim=0.
         add_zero_attn: add a new batch of zeros to the key and
                        value sequences at dim=1.
 
+        Note: if kdim and vdim are None, they will be set to embed_dim such that
+        query, key, and value have the same number of features.
+
     Examples::
 
         >>> multihead_attn = nn.MultiheadAttention(embed_dim, num_heads)
         >>> attn_output, attn_output_weights = multihead_attn(query, key, value)
     """
 
-    def __init__(self, embed_dim, num_heads, dropout=0., bias=True, add_bias_kv=False, add_zero_attn=False):
+    def __init__(self, embed_dim, num_heads, kdim=None, vdim=None, dropout=0., bias=True, add_bias_kv=False, add_zero_attn=False):
         super(MultiheadAttention, self).__init__()
         self.embed_dim = embed_dim
+        self.kdim = kdim if kdim is not None else embed_dim
+        self.vdim = vdim if vdim is not None else embed_dim
+        self.qkv_same_dim = self.kdim == embed_dim and self.vdim == embed_dim
+
         self.num_heads = num_heads
         self.dropout = dropout
         self.head_dim = embed_dim // num_heads
         assert self.head_dim * num_heads == self.embed_dim, "embed_dim must be divisible by num_heads"
 
-        self.in_proj_weight = Parameter(torch.empty(3 * embed_dim, embed_dim))
+        self.q_proj_weight = Parameter(torch.Tensor(embed_dim, embed_dim))
+        self.k_proj_weight = Parameter(torch.Tensor(embed_dim, self.kdim))
```

Comment:
when you load older models / older models' state-dicts, the model will break / compute wrong result. what are you going to do about that?

you can do what batchnorm did. see https://github.com/pytorch/pytorch/blob/master/torch/nn/modules/batchnorm.py#L16 and look at `_version = 2`,  and https://github.com/pytorch/pytorch/blob/master/torch/nn/modules/batchnorm.py#L90-L103

Reply:
Thanks for pointing this out.

add _load_from_state_dict() function to detect in_proj_weight. If in_proj_weight exists in state_dict, it will split in_proj_weight into three separate weights. I assume users know how to use state_dict() and load_state_dict() to resolve this type of version conflict.

Test on the word language model
1. load the model based by the old module, which has in_proj_weight. 
`model = torch.load("old.pt")`
2. generate state_dict. 
`state_dict = model.state_dict()`
3. create a model based on the new module, which has three separate weights. 
`new_model = new_module()`
4. map state on the new model. 
`new_model.load_state_dict(state_dict)`